File: 0493-reverse-pairs/0493-reverse-pairs.java
class Solution {
    int count = 0;

    public int reversePairs(int[] nums) {
        mergeSort(nums, 0, nums.length - 1);
        return count;
    }

    public void mergeSort(int[] nums, int left, int right) {
        if (left >= right) {
            return;
        }
        int mid = left + (right - left) / 2;

        mergeSort(nums, left, mid);
        mergeSort(nums, mid + 1, right);

        count += merge(nums, left, mid, right);
    }

    public int merge(int [] nums, int l, int m, int r){
            int i = l;
            int j = m + 1;
            int reversePairs = 0;

            while(i <= m && j <= r){
                if((long)nums[i] > 2L * nums[j]){
                    reversePairs += m - i + 1;
                    j++;
                }
                else{
                    i++;
                }
            }

            i = l;
            j = m + 1;

            ArrayList<Integer> temp = new ArrayList<>();

            while(i <= m && j <= r){
                if(nums[i] <= nums[j]){
                  temp.add(nums[i]);
                  i++;
                }
                else{
                    temp.add(nums[j]);
                    j++;
                }
            }

            while(i <= m){
                temp.add(nums[i]);
                 i++;
            }

            while(j <= r){
                temp.add(nums[j]);
                 j++;
            }

            for(int k = l; k <= r; k++){
                nums[k] = temp.get(k - l);
            }
            return reversePairs;
        }
}

File: 0493-reverse-pairs/0493-reverse-pairs.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var reversePairs = function (nums) {
    let count = 0;

    function mergeSort(left, right) {
        if (left >= right) {
            return;
        }
        let mid = Math.floor(right + (left - right) / 2);
        mergeSort(left, mid);
        mergeSort(mid + 1, right);

        count += merge(left, mid, right);
    }

    function merge(left, mid, right) {
        let i = left;
        let j = mid + 1;
        let reversePairs = 0;

        while (i <= mid && j <= right) {
            if (nums[i] > 2 * nums[j]) {
                reversePairs += mid - i + 1;
                j++;
            } else {
                i++;
            }
        }

        i = left;
        j = mid + 1;
        let temp = [];

        while (i <= mid && j <= right) {
            if (nums[i] <= nums[j]) {
                temp.push(nums[i]);
                i++;
            }
            else {
                temp.push(nums[j]);
                j++;
            }
        }

        while (i <= mid) {
            temp.push(nums[i]);
            i++;
        }

        while (j <= right) {
            temp.push(nums[j]);
            j++;
        }

        for (let k = left; k <= right; k++) {
            nums[k] = temp[k - left];
        }
        return reversePairs;
    }
    mergeSort(0, nums.length - 1);
    return count;
};